README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ClarofyCS

<!-- badges: start -->
<!-- badges: end -->

The goal of ClarofyCS is to help users interact with the Clarofy API
without the need to convert data into JSON format and packing inside GET
requests.

## Installation

You can install the released version of ClarofyCS with:

``` r
devtools::install_github("clarofy/ClarofyCS")
```

## Example

This is a basic example which shows you how to create a session with the
API:

``` r
library(ClarofyCS)
## basic example code
# url <- "SUBSTITUTE-API-URL"
# APIKEY <- "YOUR-API-KEY"
# sessionID <- CreateSession(APIKey = APIKey, url = url)
```


NEWS.md
# ClarofyCS 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/data.R
# PUBLIC IRON FLOTATION DATA SET
#'
#' Contains sample iron flotation time series data for optimisation
#'
#'
#' @format A data.frame with 24 columns and 2539 rows
#' \describe{
#' \item{XIronConcentrate}{Primary Dependent variable - Iron concentration in flotation product that we want to maximise}
#' \item{XSilicaConcentrate}{Secondary Dependent variable - Silica concentration in flotation product that we want to minimise}
#' \item{AminaFlow}{Flotation regent dose rate as a process control variable}
#' \item{StarchFlow}{Flotation regent dose rate as a process control variable}
#' \item{TimeStamp}{Time formated column}
#' \item{Upstream Forecasting Variables}{The variables that impact on the dependent variables but which may not be process control variables}
#' }
#'
#' @source Data are from a public Kaggle data set
#'
#' @examples
#' iron_flotation_data
#'
"iron_flotation_data"


R/all-in-one-functions.R
#  14.0 ALL-IN-ONE OPTIMISATION WORKFLOW ----
#' Start an all in one optimisation process
#'
#' @description
#' Start the all in one optimisation workflow
#'
#' @param url API url
#' @param APIKey API key provided
#' @param train training data set
#' @param test test data set
#' @param Independent_Variables regression variables that are used to predict dependent variable
#' @param Dependent_Variable dependent variable for which you want to predict or optimise
#' @param Upstream_Forecasting_Variables a subset of independent variables which you can't easily control
#' @param Process_Control_Variables a subset of independent variables that you want to optimise
#' @param number_clusters number of clusters to search for in upstream forecasting variables
#' @param maximise_dependent_variable logical to indicate whether to maximise or minimise the dependent variable
#' @param monitor_status logical to indicate whether session is to wait for the process to complete or return id for later retrieval
#'
#' @details
#' # Optimisation Workflow
#' - Returns the optimisation job identifier
#' - Setting monitor_status = TRUE means your current session process will
#'   be in a held state until the job is completed.
#' - Setting monitor_status = FALSE means you can continue to use your current
#'   session without waiting for the optimisation to complete.  You will be able
#'   to use the returned jobId to fetch the optimisation results at a later time.
#'
#' @examples
#'
#' \dontrun{
#'
#' all_in_one_jobId <-
#' ClarofyFinalOptimiseStart(
#'                  train = train,
#'                  test = holdout,
#'                  Dependent_Variable = Dependent_Variable,
#'                  Independent_Variables = Independent_Variables,
#'                  Process_Control_Variables = Process_Control_Variables,
#'                  Upstream_Forecasting_Variables = Upstream_Forecasting_Variables,
#'                  number_clusters = 2,
#'                  maximise_dependent_variable = FALSE,
#'                  monitor_status = TRUE,
#'                  url = url,
#'                  APIKey = APIKey)
#'
#' }
#'
#' @export
ClarofyFinalOptimiseStart <- function(url,
                                      APIKey,
                                      train,
                                      test,
                                      Dependent_Variable,
                                      Independent_Variables,
                                      Process_Control_Variables,
                                      Upstream_Forecasting_Variables,
                                      number_clusters,
                                      maximise_dependent_variable = TRUE,
                                      monitor_status = TRUE
                                      ) {

  body <- list(
    Training_Data = train
    ,Testing_Data = test
    ,Dependent_Variable = Dependent_Variable
    ,Independent_Variables = Independent_Variables
    ,Process_Control_Variables = Process_Control_Variables
    ,Upstream_Forecasting_Variables = Upstream_Forecasting_Variables
    ,Prediction_Variable_Name = "Prediction"
    ,Optimised_Prediction_Variable_Name = "Optimised_Prediction"
    ,Model_Selections = c("support vector machine")
    ,Model_Ensemble_Weights = c(1)
    ,TimeStamp_Variable_Name = "TimeStamp"
    ,Number_Of_Clusters = number_clusters
    ,Buffer = 3
    ,Weights = NULL
    ,Feature_Scaling_Method = "z-score normalization"
    ,Maximise_Model_To_Optimise = maximise_dependent_variable
    ,Export_Model_Name = stringr::str_glue("svm_", Sys.Date())
  )

  # Create JSON for Session Jobs
  jsonlite::write_json(x = body, path = "./allinone.json")
#
  # CREATE SESSION
  path <- '/Session/CreateSession'
  bodyPost <- "{}"
  raw.result <- httr::POST(url = url, path = path, body = bodyPost, encode = 'json', httr::add_headers(APIKey = APIKey))
  Data <- rawToChar(raw.result$content)
  print(paste0("[",Sys.time(),"] Session ID: ",Data))
  id <- Data


  # SPAWN SESSION
  path <- '/Session/Interlate_All_In_One_Optimisation_Start'
  bodyPost <- jsonlite::read_json(path="./allinone.json")
  raw.result <- httr::POST(url = url, path = path, body = bodyPost, encode = 'json', httr::add_headers(SessionKey = id))
  Data <- rawToChar(raw.result$content)
  jobId <- Data
  print(paste0("[",Sys.time(),"] Load JSON Result: ",jobId))

  # Save copy of session and job to restore if required
  restore_last_session <- list(
    id = id,
    jobId = jobId
  )

  saveRDS(restore_last_session, file = "restore_last_session")

  if (monitor_status) {

    repeat {
      path <- paste0('/Session/Interlate_Conditional_Optimisation_Progress/', jobId)
      raw.result <- httr::GET(url = url, path = path)
      progress <- rawToChar(raw.result$content)
      print(paste0("[",Sys.time(),"] ",progress))
      if (grepl("Completed", progress)){
        break;
      }
      Sys.sleep(60)

    }

    return(jobId)

  } else {

    return(jobId)

  }

}


#' Check status of all-in-one optimisation workflow
#'
#' @description
#' Returns status of all-in-one optimisation workflow
#'
#' @param url API url
#' @param id Session id created by [CreateSession]
#' @param jobId job id for temporary processes run during a session
#'
#' @details
#' # Optimisation Workflow
#' - Returns optimised predictions for the dependent variable using the optimal
#'   process control variables
#'
#' @examples
#'
#' \dontrun{
#'
#' all_in_one_jobStatus <-
#' ClarofyFinalOptimiseStatus(url = url, id = sessionID, jobId = all_in_one_jobId)
#'
#' }
#'
#' @export
ClarofyFinalOptimiseStatus <- function (url, id, jobId) {

  print(paste0("[",Sys.time(),"] Fetching Job Final Status"))
  path <- paste0('/Session/Interlate_All_In_One_Optimisation_Results/', jobId)
  raw.result <- httr::GET(url = url, path = path, httr::add_headers(SessionKey = id))
  results <- rawToChar(raw.result$content)
  print(paste0("[",Sys.time(),"] Job Status: ",results))

  return(results)

}

#' Get the productivity groups
#'
#' @description
#' Returns the productivity groups from the all-in-one optimisation workflow
#'
#' @param url API url
#' @param id Session id created by [CreateSession]
#'
#' @details
#' # Optimisation Workflow
#' - Returns the producitivity groups from the all-in-one optimisation workflow
#'
#' @examples
#'
#' \dontrun{
#'
#' productivityGroups <- ClarofyGetProductivityGroups(url = url, id = sessionID)
#'
#' }
#'
#' @export
ClarofyGetProductivityGroups <- function(url, id) {

  # Load Productivity Groups
  print(paste0("[",Sys.time(),"] Getting Productivity Groups..."))
  path <- '/Session/SessionData/ProductivityGroups'
  raw.result <- httr::GET(url = url, path = path, httr::add_headers(SessionKey = id))
  productivityGroups <- jsonlite::fromJSON(rawToChar(raw.result$content))

  # Print out the productivity groups to Viewer
  print(productivityGroups[[2]] %>% DT::datatable(options = list(
    initComplete = DT::JS(
      "function(settings, json) {",
      "$(this.api().table().header()).css({'background-color': '#251b29', 'color': '#fff'});",
      "}")
  )))

  return(productivityGroups)
}

#' Get the optimised results
#'
#' @description
#' Returns the optimised results from the all-in-one optimisation workflow
#'
#' @param url API url
#' @param id Session id created by [CreateSession]
#'
#' @details
#' # Optimisation Workflow
#' - Returns the optimised results from the all-in-one optimisation workflow
#' - Returns a time series plot of the predictions
#'
#' @examples
#'
#' \dontrun{
#'
#' FinalOptimisedPredictions <-
#' ClarofyGetFinalOptimisedResults(url = url, id = sessionID)
#'
#' }
#'
#' @export
ClarofyGetFinalOptimisedResults <- function(url, id) {

    print(paste0("[",Sys.time(),"] Getting Optimised Predictions..."))
    path <- '/Session/SessionData/OptimisedPredictions'
    raw.result <- httr::GET(url = url, path = path, httr::add_headers(SessionKey = id))
    optimisedPredictions <- jsonlite::fromJSON(rawToChar(raw.result$content))

    # ** Visualise Optimised Prediction Vs Actual KPI ----
    print(optimisedPredictions %>%
            dplyr::select(TimeStamp, Dependent_Variable, Prediction, Optimised_Prediction, type) %>%
            dplyr::mutate(TimeStamp = lubridate::ymd_hms(TimeStamp)) %>%
            tidyr::pivot_longer(-c(TimeStamp, type)) %>%
            timetk::plot_time_series(.date_var = TimeStamp,
                             .value = value,
                             .color_var = name,
                             .line_alpha = 0.4,
                             .facet_vars = type,
                             .smooth = FALSE,
                             .title = "Modeled vs Optimised Predictions for Training and Holdout Dataset"))

    return(optimisedPredictions)

}

#' Get the sensitivity analysis from the all-in-one optimisation workflow
#'
#' @description
#' Returns the sensitivity estimates from the all-in-one optimisation workflow
#'
#' @param url API url
#' @param id Session id created by [CreateSession]
#'
#' @details
#' # Optimisation Workflow
#' - Returns the sensitivity estimates from the all-in-one optimisation workflow.
#' - The values of each process control variable in the recommendation results is
#'   used to replace the values of the identical process control variable in the
#'   original dataset.
#' - The modified dataset is passed into the trained Prediction model and the
#'   predicted value for the dependent variable is stored.
#' - The percentage contribution of each individual optimised process control
#'   variable to the overall optimised prediction is returned as sensitivity
#'   estimates.
#'
#' @examples
#'
#' \dontrun{
#'
#' sensitivityAnalysis <-
#' ClarofyGetSensitivityAnalysis(url = url, id = sessionID)
#'
#' }
#'
#' @export
ClarofyGetSensitivityAnalysis <- function(url, id) {

  # Sensitivity Analysis ----
  print(paste0("[",Sys.time(),"] Getting Sensitivity Analysis..."))
  path <- '/Session/SessionData/SensitivityAnalysis'
  raw.result <- httr::GET(url = url, path = path, httr::add_headers(SessionKey = id))
  sensitivityAnalysis <- jsonlite::fromJSON(rawToChar(raw.result$content))

  return(sensitivityAnalysis)

}




R/Clarofy-Demo-Iron-Flotation.R
# Test ClarofyCS API Client Side Functions ----
# This script can be run in conjunction with an active Clarofy API Key and
# valid API url provided to each user.

{
library(ClarofyCS)
library(DT)
library(rsample)
library(timetk)
library(dplyr)
library(tidyr)
library(ggplot2)
library(yardstick)
library(janitor)
library(lubridate)
library(stringr)


url <- "enter-api-url"
APIKey <- "enter-api-key"
}



# 1.0 CREATE SESSION FUNCTION ----

sessionID <- CreateSession(APIKey = APIKey, url = url)
sessionID

# 2.0 Set Up Global Input Definitions ----
{
  Dependent_Variable = c('XIronConcentrate')
  
  Upstream_Forecasting_Variables <- c('XIronFeed'
                                      ,'XSilicaFeed'
                                      ,'OrePulpFlow'
                                      ,'OrePulppH'
                                      ,'OrePulpDensity'
                                      ,'FlotationColumn01AirFlow'
                                      ,'FlotationColumn02AirFlow'
                                      ,'FlotationColumn03AirFlow'
                                      ,'FlotationColumn04AirFlow'
                                      ,'FlotationColumn05AirFlow'
                                      ,'FlotationColumn06AirFlow'
                                      ,'FlotationColumn07AirFlow'
                                      ,'FlotationColumn01Level'
                                      ,'FlotationColumn02Level'
                                      ,'FlotationColumn03Level'
                                      ,'FlotationColumn04Level'
                                      ,'FlotationColumn05Level'
                                      ,'FlotationColumn06Level'
                                      ,'FlotationColumn07Level')
  
  Process_Control_Variables <- c('StarchFlow'
                                 ,'AminaFlow')
  
  
  
  Independent_Variables <- c(Upstream_Forecasting_Variables, Process_Control_Variables)
}

# 3.0 IMPORT/LOAD DATA ----

databaseData <- iron_flotation_data

databaseData %>% glimpse()


# Data Prep ----
{
  # Data Prep ----
  Data_prep <- databaseData %>%
    mutate(TimeStamp = ymd_hms(TimeStamp)) %>%
    arrange(TimeStamp)
  
  splits <- initial_time_split(Data_prep, prop = 0.8)
  
  splits %>%
    tk_time_series_cv_plan() %>%
    plot_time_series_cv_plan(TimeStamp, !! sym(Dependent_Variable))
  
  train <- training(splits)
  holdout <- testing(splits)
}

# 3.0 LOAD DATAFRAME INTO SESSION FUNCTION ----

# Test Load Dataframe and Get Load Status ----
loadStatus <- LoadDataFrameIntoSession(df      = databaseData, 
                                       url     = url, 
                                       id      = sessionID)  

loadStatus


# GET LOADED DATA FUNCTION ----

getData <- GetData(id = sessionID, url = url)

getData %>% glimpse()


# 4.0 GET SUMMARY OF LOADED DATA ----

Summary <- GetSummaryOfSessionData(url  = url, 
                                   id   = sessionID)


# 5.0 LOAD TRAINING DATA INTO SESSION ----

train_load_status <- LoadTrainingDataIntoSession(train = train, 
                                                 url = url, 
                                                 id = sessionID)

train_load_status


# 6.0 LOAD TESTING DATA INTO SESSION ----

test_load_status <- LoadTestingDataIntoSession(test = holdout, 
                                               url = url, 
                                               id = sessionID)

test_load_status


# 7.0 CLEAN ALL VARIABLES USING DATA DENSITY CLEANSING OPERATION ----


Data_Cleaned <- CleanseAllTestingData(Dependent_Variable = Dependent_Variable,
                                      Independent_Variables = Independent_Variables,
                                      url = url, id = sessionID)

Data_Cleaned %>% glimpse()

# 8.0 COMPARE CLEANED DATA WITH RAW TRAINING DATA ----
{
raw_pivot <- train %>%
  pivot_longer(cols = contains("Column"),
               names_to = "metric",
               values_to = "value") %>%
  mutate(type = "raw")

clean_pivot <- Data_Cleaned %>%
  pivot_longer(cols = contains("Column"),
               names_to = "metric",
               values_to = "value") %>%
  mutate(type = "filtered")

raw_pivot %>%
  bind_rows(clean_pivot) %>%
  ggplot(aes(value, fill = type)) +
  geom_density(alpha = 0.5) +
  facet_wrap(~ metric, scales = "free") +
  theme_minimal()


data_compare <- Data_Cleaned %>%
  left_join(databaseData, by = c("TimeStamp" = "TimeStamp")) %>%
  select(TimeStamp, starts_with(Dependent_Variable)) %>%
  rename(Cleaned_KPI = str_glue("{Dependent_Variable}.x"),
         Original_KPI = str_glue("{Dependent_Variable}.y"))

data_compare %>%
  pivot_longer(-TimeStamp) %>%
  plot_time_series(.date_var = TimeStamp,
                   .value = value,
                   .color_var = name,
                   .line_alpha = 0.4,
                   # .facet_vars = name,
                   .smooth = FALSE,
                   .title = "Comparison of Cleaned and Original Datasets")

}

# 9.0  TRAIN A MODEL ----

Model_Trainer_Output <- ClarofyFit(Dependent_Variable = Dependent_Variable,
                                   Independent_Variables = Independent_Variables,
                                   url = url,
                                   id = sessionID)

Model_Trainer_Output

# 10.0 MAKE PREDICTIONS WITH MODEL ----

# Training Data predictions - in sample
Data_Predictions_train <- ClarofyPredict(Dependent_Variable = Dependent_Variable,
                                         url = url, id = sessionID, df = train)


# 11.0 PLOT PREDICTIONS VS ACTUAL OBSERVATIONS ----
{
  custom_metrics_reg <- yardstick::metric_set(yardstick::mae, yardstick::rmse, yardstick::rsq)
  
  print('Custom metrics function')
  galaxy_model_metrics_train <- Data_Predictions_train %>%
    custom_metrics_reg(!! sym(Dependent_Variable), "Prediction")
  
  print('Custom metrics function')
  galaxy_model_metrics_train %>%
    clean_names(case = 'title') %>%
    select(-Estimator) %>%
    mutate(Metric = toupper(Metric),
           Estimate = round(Estimate, digits = 3)) %>%
    datatable(options = list(
      initComplete = JS(
        "function(settings, json) {",
        "$(this.api().table().header()).css({'background-color': '#251b29', 'color': '#fff'});",
        "}")
    ))
  
  Data_Predictions_train %>%
    filter(!!sym(Dependent_Variable) > 0.025) %>%
    ggplot(aes(!! sym(Dependent_Variable),Prediction)) +
    geom_point(alpha = 0.5) +
    geom_abline(color = 'blue', linetype = 2) +
    tune::coord_obs_pred() +
    labs(title = "R squared plot for evaluating model performance - In Sample",
         subtitle = "Actual vs Model",
         x = 'Actual KPI', y = 'Model Predicted KPI')
}
{
Data_Predictions_train %>%
  select(TimeStamp, contains(Dependent_Variable), Prediction) %>%
  filter(Prediction > 0.02) %>%
  pivot_longer(-TimeStamp) %>%
  plot_time_series(.date_var = TimeStamp,
                   .value = value,
                   .color_var = name,
                   .smooth = FALSE,
                   .title = "Model Performance on Training Data (In Sample)",
                   .line_alpha = 0.6)
}

# Testing Data predictions - in sample
Data_Predictions_test <- ClarofyPredict(Dependent_Variable = Dependent_Variable,
                                         url = url, id = sessionID, df = holdout)


{
  custom_metrics_reg <- yardstick::metric_set(yardstick::mae, yardstick::rmse, yardstick::rsq)
  
  print('Custom metrics function')
  galaxy_model_metrics_test <- Data_Predictions_test %>%
    custom_metrics_reg(!! sym(Dependent_Variable), "Prediction")
  
  print('Custom metrics function')
  galaxy_model_metrics_test %>%
    clean_names(case = 'title') %>%
    select(-Estimator) %>%
    mutate(Metric = toupper(Metric),
           Estimate = round(Estimate, digits = 3)) %>%
    datatable(options = list(
      initComplete = JS(
        "function(settings, json) {",
        "$(this.api().table().header()).css({'background-color': '#251b29', 'color': '#fff'});",
        "}")
    ))
  
  Data_Predictions_test %>%
    filter(!!sym(Dependent_Variable) > 0.025) %>%
    ggplot(aes(!! sym(Dependent_Variable),Prediction)) +
    geom_point(alpha = 0.5) +
    geom_abline(color = 'blue', linetype = 2) +
    tune::coord_obs_pred() +
    labs(title = "R squared plot for evaluating model performance - Out of Sample",
         subtitle = "Actual vs Model",
         x = 'Actual KPI', y = 'Model Predicted KPI')
}
{
Data_Predictions_test %>%
  select(TimeStamp, contains(Dependent_Variable), Prediction) %>%
  filter(Prediction > 0.02) %>%
  pivot_longer(-TimeStamp) %>%
  plot_time_series(.date_var = TimeStamp,
                   .value = value,
                   .color_var = name,
                   .smooth = FALSE,
                   .title = "Model Performance on Training Data (Out of Sample)",
                   .line_alpha = 0.6)
}
# 12.0 GENERATE TARGET FOR OPTIMISATION ----
ProductivityGroups <- ClarofyGroup(Dependent_Variable = Dependent_Variable,
                                   Upstream_Forecasting_Variables = Upstream_Forecasting_Variables,
                                   n = 10,
                                   url = url,
                                   id = sessionID)
{
# For demonstration of an iteration step we select one productivity grouping of
# upstream forecasting variables.  This could also be a time weighted average for
# # real time recommendations using actual conditions.
# 
# ProductivityGroups$Summary_Of_Clusters %>% 
#   select(Upstream_Forecasting_Variables) %>% as_tibble()

# Run Optimisation for Each Productivity Group

# Filter each productivity group and run optimisations:
prod_group <- 6
centroid <- ProductivityGroups$Summary_Of_Clusters[prod_group,] %>%
  select(all_of(Upstream_Forecasting_Variables))

centroid

}

# 13.0 SINGLE OPTIMISATION ----
# The optimisation function must be set to either minimise or maximise the KPI depending on need
# The returned object includes the optimised process control variables as defined in step 2.0
# Use the recommended process control setpoints and actual upstream forecasting variables to
# observe the expected difference between the actual and optimal KPI.

jobId <- ClarofyOptimiseStartJob(Dependent_Variable = Dependent_Variable,
                         Upstream_Forecasting_Variables = Upstream_Forecasting_Variables,
                         Process_Control_Variables = Process_Control_Variables,
                         upstream_values = centroid, 
                         maximise_dependent_variable = TRUE, 
                         monitor_status = TRUE, 
                         url = url, id = sessionID)

jobStatus <- ClarofyOptimiseJobStatus(url = url, id = sessionID, jobId = jobId)

jobStatus

ProcessControlRecommendations <- ClarofyRecommend(url = url, id = sessionID)

OptimisedPredictions <- ClarofyOptimisedPredictions(url = url, id = sessionID)

OptimisedPredictions %>% glimpse()

ClarofyPlotOptimisationRun(Dependent_Variable = Dependent_Variable, ncol = 3,
                           url = url, id = sessionID)

# 14.0 All IN ONE OPTIMISATION ----
all_in_one_jobId <- ClarofyCS::ClarofyFinalOptimiseStart(url = url, 
                                           APIKey = APIKey, 
                                           train = train, 
                                           test = holdout, 
                                           Dependent_Variable = Dependent_Variable, 
                                           Independent_Variables = Independent_Variables, 
                                           Process_Control_Variables = Process_Control_Variables,
                                           Upstream_Forecasting_Variables = Upstream_Forecasting_Variables, 
                                           number_clusters = 2, 
                                           maximise_dependent_variable = FALSE, 
                                           monitor_status = TRUE)



# * Restore point ----
# Restore points are created by the client services package during the long running optimisation processes.
# Uncomment and run code below if RStudio session is aborted or closed while session and job are running
# readRDS(file = "restore_last_session.rds")


all_in_one_jobStatus <- ClarofyFinalOptimiseStatus(url = url, id = sessionID, jobId = all_in_one_jobId)

productivityGroups <- ClarofyGetProductivityGroups(url = url, id = sessionID)

FinalOptimisedPredictions <- ClarofyGetFinalOptimisedResults(url = url, id = sessionID)

FinalOptimisedPredictions %>% glimpse()

sensitivityAnalysis <- ClarofyGetSensitivityAnalysis(url = url, id = sessionID)

sensitivityAnalysis %>% glimpse()

# 15.0 FINISH SESSION ----

ClarofyEndSession(url = url, id = sessionID)
